[WIP] Panoptic segmentation

vniclas opened the pull request:
Hi, 
this PR adds the [EfficientPS](http://panoptic.cs.uni-freiburg.de/) network. The original repo can be found [here](https://github.com/DeepSceneSeg/EfficientPS).

**Todos:**
- [ ] Upload pre-trained models to OpenDR server and adjust the URLs in [efficient_ps_learner.py](https://github.com/tasostefas/opendr_internal/blob/panoptic-segmentation/src/opendr/perception/panoptic_segmentation/efficient_ps/efficient_ps_learner.py). _Where are the instructions to upload something to the server?_
- [ ] Add unit tests

**Known issues:**
- 3rd party dependencies require pytorch to be installed first since they attempt to load `torch` in their `setup.py`

Pavlos-Tosidis commented:
> Hi,
> this PR adds the [EfficientPS](http://panoptic.cs.uni-freiburg.de/) network. The original repo can be found [here](https://github.com/DeepSceneSeg/EfficientPS).
> 
> **Todos:**
> 
>     * [ ]  Upload pre-trained models to OpenDR server and adjust the URLs in [efficient_ps_learner.py](https://github.com/tasostefas/opendr_internal/blob/panoptic-segmentation/src/opendr/perception/panoptic_segmentation/efficient_ps/efficient_ps_learner.py). _Where are the instructions to upload something to the server?_

You should upload the files to OpenDR's owncloud, under the FTP server material, with the desired structure. When you upload the files, send me an email and I will take care of uploading the files to the FTP server

passalis commented:
@vniclas testing is now available again. So, when you are ready, you can use the "test sources" label to run the style checks and "test tools" to run the tool tests.
Please let us know when this PR is ready for review.

stefaniapedrazzi commented:
Is there a particular reason why the `mmcv` library is used for the example (at least the one is supposed to be added in D7.1) instead of the opencv library that is already a dependency of the opendr toolkit?
We should keep the number of dependencies as low as possible and reuse the same as much as possible.

vniclas commented:
> Is there a particular reason why the `mmcv` library is used for the example (at least the one is supposed to be added in D7.1) instead of the opencv library that is already a dependency of the opendr toolkit?
> We should keep the number of dependencies as low as possible and reuse the same as much as possible.

The examples and test files have been changed to `cv2` for opening image files. However, we cannot completely remove `mmcv` as a dependency since the underlying base code relies on the library.

stefaniapedrazzi commented:
I tried to run the sample example but I'm getting some errors:
```
>>> from opendr.perception.panoptic_segmentation.datasets import CityscapesDataset
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/stefi/develop/opendr_internal/src/opendr/perception/panoptic_segmentation/datasets/__init__.py", line 2, in <module>
    from opendr.perception.panoptic_segmentation.datasets.cityscapes import CityscapesDataset
  File "/home/stefi/develop/opendr_internal/src/opendr/perception/panoptic_segmentation/datasets/cityscapes.py", line 28, in <module>
    from mmdet.datasets import CityscapesDataset as MmdetCityscapesDataset
ModuleNotFoundError: No module named 'mmdet'
```

It seems that some dependencies are not correctly listed.

vniclas commented:
As written in the _known issues_ section of the PR, the dependencies have not been completely sorted yet since the base repositories assume a manual installation of `pytorch` before a successful installation. Just using `pip install -r requirements.txt` does not work.
If you have any idea how to integrate a 2 step installation process in the OpenDR environment, please let me know.

stefaniapedrazzi commented:
Ok, so the example added in D7.1 it currently doesn't work and cannot be tested.

For installation you can simply add a bash script that will be then called by the main installation system.
However, the OpenDR toolkit should use a single version of pytorch for the whole library.
So the installed version (manually or whatever) should work with all the provided tools and all the methods requirements should be adjusted to use the exact same version.

But what does exactly requires the third party dependency?
A specific version, specific installation setup, specific installation path?

stefaniapedrazzi commented:
>  3rd party dependencies require pytorch to be installed first since they attempt to load torch in their setup.py

First with respect of what?

vniclas commented:
I checked all 3rd party tools used in this PR. The minimum common requirement for pytorch is `torch>=1.4.0`, however the pre-trained models that have been made available work only with `torch==1.7.x`. I have not found any minimum version number specified by the OpenDR toolkit. Could you please point me to some documentation (if it exists)? Are there more restrictions induced by other python libraries or even the base interpreter?

Basically, the problem is that both [inplace_abn](https://github.com/mapillary/inplace_abn) and [mmdet](https://github.com/open-mmlab/mmdetection) attempt to load `torch` in their respective `setup.py` files for building the libraries, i.e., the installation will fail if pytorch has not been installed beforehand.

stefaniapedrazzi commented:
> I checked all 3rd party tools used in this PR. The minimum common requirement for pytorch is `torch>=1.4.0`, however the pre-trained models that have been made available work only with `torch==1.7.x`. I have not found any minimum version number specified by the OpenDR toolkit. Could you please point me to some documentation (if it exists)? Are there more restrictions induced by other python libraries or even the base interpreter?

The documentation where a first set of dependencies is specified is D2.1:
https://cicloud.csd.auth.gr/owncloud/apps/files/?dir=/OpenDR/Documents/WP2/Deliverables&fileid=48
Here we specified that we would use PyTorch 1.5.0.
Note that if needed, the pytorch version can be changed but it should be changed in all the tools and all the tools needs to support it.

> Basically, the problem is that both inplace_abn and mmdet attempt to load torch in their respective setup.py files for building the libraries, i.e., the installation will fail if pytorch has not been installed beforehand.

I would say that the issues is rather in these libraries than in the pip installation system.
But in any case, pytorch is a main requirements of the OpenDR toolkit and can be installed in any case just before the specific tools dependencies, for example by adding it here:
https://github.com/tasostefas/opendr_internal/blob/94f6bc0d9d5f2ca5e68d79394754363ec17600f9/dependencies/install.sh#L13
Would it work in this case?

vniclas commented:
Thanks, @stefaniapedrazzi! That helped a lot. Sorry that I was not aware of this specification since it is not mentioned in the repo's wiki and I just joined this project one month ago. I added `torch==1.5.1` to the `install.sh` script.

Do you have any idea why clang fails on ubuntu-20, whereas it passes on ubuntu-18? I formatted the files with clang-format-9 on an ubuntu-20 machine.

stefaniapedrazzi commented:
>  I added torch==1.5.1 to the install.sh script

I just checked the code and all the already merged methods are using `torch` 1.7.1.
So if possible your script should also install 1.7.1. Otherwise we will have to find a version working for all the tools or adjust some of them.

About the error with clang, it may be that a different version is installed depending on the system.
This can be checked and fixed.
But it seems that your are including lot of files.
Did you copy the files in the `algorithm` folder from a third party or did you write all of them specifically from the OpenDR toolkit?
If they are copied from another source then they should not follow the OpenDR style and the clang-format test could be skipped for it.
Even better, if they are stored on another GitHub repo, we should linkn to them as a sumodule instead of include the files directly in this repo.

vniclas commented:
Thanks for your reply. I changed the installed pytorch version to 1.7.1 and replaced the code in the `algorithm` folder with a submodule linking to the original repository.
Currently, I am facing two issues:
- The current test does not check out submodules and hence fails.
- The dependency [inplace_abn](https://github.com/mapillary/inplace_abn) cannot be installed in the test environment due to `OSError: CUDA_HOME environment variable is not set. Please set it to your CUDA install root.`.

Can we adapt the workflow file accordingly?

stefaniapedrazzi commented:
> Thanks for your reply. I changed the installed pytorch version to 1.7.1 and replaced the code in the `algorithm` folder with a submodule linking to the original repository.
> Currently, I am facing two issues:
> 
>     * The current test does not check out submodules and hence fails.
> 
>     * The dependency [inplace_abn](https://github.com/mapillary/inplace_abn) cannot be installed in the test environment due to `OSError: CUDA_HOME environment variable is not set. Please set it to your CUDA install root.`.
> 
> 
> Can we adapt the workflow file accordingly?

I updated the workflow so that now it also downloads and initializes the submodules.
I also skipped the style check on the new submodule.

About CUDA, currently the CI doesn't have NVIDIA drivers installed.
It would be better if tests could run without it.
Otherwise we will have to change the workflow installation. But please go ahead inspecting this. I will be happy to review your proposal to install CUDA on the CI.

vniclas commented:
Update concerning CUDA installation in CI

I managed to install CUDA using the [cuda-toolkit package](https://github.com/Jimver/cuda-toolkit) such that pytorch returns `torch.has_cuda: True`. However, this doesn't suffice to install `inplace_abn`. Alternatively, the dependency can also be compiled for CPUs only if we install `torch==1.7.1+cpu` in CI.

The problem though is that EfficientPS is based on an older release of [mmdetection](https://github.com/open-mmlab/mmdetection) that only compiles if `torch.cuda.is_available() == True`. Obviously, this is not the case on our CI system since there are no GPU nodes. In a more recent release of mmdetection, a CPU-only compilation seems to be possible if I understand the `setup.py` correctly. I talked to the original author of EfficientPS and there will not be any upgrade to the latest version of mmdetection in the near future.

Basically, this leaves us with two options:
1) Skip the tests for EfficientPS in CI.
2) Copy the source code of EfficientPS to our repo and try to update it ourselves. This could take quite some time though.

stefaniapedrazzi commented:
Seems ok to make an exception and skip the tests in the CI.

vniclas commented:
In order to skip the tests for GPU-only packages in CI, I propose extending the installation and testing procedure by adding a flag to the respective make commands to enable a CPU-only mode.

MWE:
`make install_runtime_dependencies mode=cpu`

```
mode=gpu

install_runtime_dependencies:
	@+echo "#"; echo "# * Install Runtime Dependencies *"; echo "#"
ifeq (${mode}, cpu)
	@+cd dependencies; ./install.sh runtime cpu
else
	@+cd dependencies; ./install.sh runtime
endif
```

Then, we can use this flag to blacklist certain packages and exclude their dependencies from being installed and their tests from being run. 

Let me know what you think @stefaniapedrazzi, then I can start with making a PR for such a feature.

stefaniapedrazzi commented:
Yes, the idea is good to have two different installation procedure for CPU-based and GPU-based system (that will install everything) but I think it is much better to add an environment variable than an additional Makefile compilation parameter.
For example name `OPENDR_DEVICE`.
So then it is not even needed to passed as argument to the installation script because it is already set in the environment.

vniclas commented:
This PR is now ready for review again. Some comments from my side:

- The tool requires a GPU due to the underlying base repository (older version of [mmdet](https://github.com/open-mmlab/mmdetection)
- CI tests are skipped as discussed before
- I commented on the `ImageWithFilename` class suggesting to merge it with `Image` or move to the same engine file.

vniclas commented on docs/reference/efficient-ps.md:
```diff
@@ -0,0 +1,163 @@
+- **mode**: *str*
+  Valid options are 'model' to download pre-trained model weights and 'test_data' to run the unit tests.
+- **trained_on**: *str*
+  Specifies which model weights to download. Pre-trained models are available for the Cityscapes and KITTI panoptic segmentation datasets.\
```
ToDo @vniclas : remove the backslash

vniclas commented on src/opendr/perception/panoptic_segmentation/datasets/utils.py:
```diff
@@ -0,0 +1,34 @@
+from opendr.engine.data import Image as BaseImage
+
+
+class Image(BaseImage):
```
This could be merged with the `opendr.engine.data.Image` class or renamed to `ImageWithFilename` and moved to the same engine file.

passalis replied:
Why do we need the filename of the image? Is this needed during training or inference?

vniclas replied:
Refactored the code to remove the need for an additional `ImageWithFilename` implementation.

passalis commented on src/opendr/perception/panoptic_segmentation/efficient_ps/efficient_ps_learner.py:
```diff
@@ -0,0 +1,521 @@
+        else:
+            return results
+
+    def save(self, path: str) -> bool:
```
There is a specific OpenDR format for saving/loading models that encapsulates other existing formats in order to provide a uniform interface. This is specified in D2.1 and the already integrated tools provide an example of how to load/save using this format. Essentially, you need to create a folder with the name of your model and include a json metadata file that describes the files that the model needs. Then, you can save the model files inside this folder. Please let me know if you need additional information.

vniclas replied:
Sorry about that. I changed it according to D2.1and also checked other tools' implementations. I do not think that all of them really follow D2.1. It feels also a bit weird that _name_ cannot be specified by the user. I further guess _x_ does not apply in my case.
Perhaps, I am misinterpreting the specification.

**From D2.1**:
The structure of files that are stored for a trained model is the following (assuming that the model is called “name”:):
• “/name/name.json”, which contains the necessary metadata for the model,
• “/name/model_x.pth”, or “/name/model_x.h5tf”, or “/name/model_x.tf”, which stores the actual model. Note that multiple such model files might exist for complex models (this is specified in the metadata JSON file (x corresponds to the id of each model)

vniclas replied:
Should we also update the files uploaded to the OpenDR server or just download the `.pth` file from there and create the `.json` when downloading the model weights?

passalis replied:
Both options sounds fine.

passalis commented on src/opendr/perception/panoptic_segmentation/efficient_ps/efficient_ps_learner.py:
```diff
@@ -0,0 +1,521 @@
+            print(msg)
+
+        shutil.rmtree(self.temp_path)
+        return results
```
Are these results a dictionary (eval should return a dictionary according to specs)?

vniclas replied:
Yes, as specified in the function's type hints, the returned `results` are a dictionary comprising panoptic, segmentation, and recognition qualities.

passalis commented on src/opendr/perception/panoptic_segmentation/efficient_ps/efficient_ps_learner.py:
```diff
@@ -0,0 +1,521 @@
+            runner.register_hook(EvalHook(val_dataloader, interval=1, metric=['panoptic']))
+
+        runner.run(dataloaders, self._cfg.workflow, self.iters)
+        self._is_model_trained = True
```
I think *fit()* is also expected to return a dictionary with the training statistics.

vniclas replied:
I addressed this by returning the training (losses) and validation (metrics) statistics. The description in D2.1 does not directly translate to my use case.

passalis reviewed (commented):
Thank you for your tool! This is an overall nice implementation. I have made a few comments mainly related to the way the Learners are supposed to behave based on the specification of the toolkit. Let us know if you have any questions on these or you need clarifications/help.

vniclas commented:
**Disclaimer:** The tests are failing right now. I will solve this after Oct. 4.

vniclas commented on src/opendr/perception/semantic_segmentation/bisenet/bisenet_learner.py:
```diff
@@ -189,7 +189,7 @@ def infer(self, img):
         self.model.eval()
         predict = self.model(image).argmax(dim=1).squeeze().cpu()
-        heatmap = Heatmap(predict)
+        heatmap = Heatmap(predict.numpy())
```
**Note:** Fixing this even though it is not in my package. The PR of `bisenet` did not implement type checking for the `Heatmap` class although only Numpy arrays were expected according to the original comments in the code.

ad-daniel commented on tests/test_pep8.py:
```diff
@@ -31,6 +31,7 @@
     '.git',
     'dependencies',
     'lib',
+    'src/opendr/perception/panoptic_segmentation/efficient_ps/algorithm/EfficientPS',
```
In the ReadMe you mentioned that the code was modified to comply with pep8, so it shouldn't be skipped

vniclas replied:
Sorry, this has to be removed from the readme. Forgot that I mentioned it there before adding this line as asked by Stefania.

ad-daniel commented on src/opendr/perception/panoptic_segmentation/efficient_ps/example_usage.py:
```diff
@@ -0,0 +1,93 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
The example/demo should be placed in the `projects` directory ([details](https://github.com/tasostefas/opendr_internal/wiki/Adding-New-Tool#example))

ad-daniel reviewed (changes requested):
Looks good overall, thank you! Please double check the default parameters, as I did it very quickly.

passalis reviewed (approved):
Thank you!
